pkgs/applications/science/logic/elan/0001-dynamically-patchelf-binaries.patch

```diff
@@ -22,13 +22,13 @@ index c51e76d..ae8159e 100644
 +            .output();
 +    }
 +
-+    if dest_path.extension() == Some(::std::ffi::OsStr::new("lld")) {
++    if dest_path.file_name() == Some(::std::ffi::OsStr::new("leanc")) {
 +        use std::os::unix::fs::PermissionsExt;
 +        let new_path = dest_path.with_extension("orig");
 +        ::std::fs::rename(dest_path, &new_path)?;
 +        ::std::fs::write(dest_path, format!(r#"#! @shell@
-+exec -a "$0" {} "$@" --dynamic-linker=@dynamicLinker@
-+"#, new_path.to_str().unwrap()))?;
++LEAN_CC=@cc@ exec -a "$0" {} "$@" -L {}/lib  # use bundled libraries, but not bundled compiler that doesn't know about NIX_LDFLAGS
```

Comment:
Ideally we should only override the `LEAN_CC` environment variable if it is not already set.

Also the `-L {}/lib` rubs me the wrong way.  Shouldn't `leanc` add this by itself if `LEAN_CC` is set?

Reply:
> Also the `-L {}/lib` rubs me the wrong way. Shouldn't `leanc` add this by itself if `LEAN_CC` is set?

Yeah, there is some tension between "using leanc with LEAN_CC for compiling Lean code" and "using leanc --print-ldflags for linking external code against Lean libraries". I'm not sure if I'm emotionally ready to introduce yet another CMake variable to distinguish between these two cases and provide appropriate flags in each case.

Reply:
So just to name the alternative solution, we could just stick NIX_CFLAGS_COMPILER and NIX_LDFLAGS in there and we would probably already be covering 99% of the libraries out there without fully emulating cc-wrapper. Then we wouldn't have to worry about discrepancies with upstream.

Reply:
In the spirit of "work the same on all platforms". I think we should eventually add the -L ../lib flag in leanc.  (Otherwise setting LEAN_CC works on nixos, but not on other platforms, right?)

But we need to add the -L ../lib flag in nixpkgs anyhow for backwards compatibility.

Reply:
> So just to name the alternative solution, we could just stick NIX_CFLAGS_COMPILE and NIX_LDFLAGS in there and we would probably already be covering 99% of the libraries out there without fully emulating cc-wrapper.

I'm not sure what you mean, those two variables are not set on nixos:
```
$ echo $NIX_CFLAGS_COMPILE $NIX_LDFLAGS

```

And even if they're set, they don't include the include path to the libc headers:
```
$ nix-shell -p zlib --command 'echo $NIX_CFLAGS_COMPILE $NIX_LDFLAGS'
-frandom-seed=f1ilflm2wm -isystem /nix/store/kva76kxzfljikns3yxrxyn76m9gdmqc4-zlib-1.2.11-dev/include -isystem /nix/store/kva76kxzfljikns3yxrxyn76m9gdmqc4-zlib-1.2.11-dev/include -rpath /nix/store/f1ilflm2wmkcq90b0bljcfs9vz0h41bb-shell/lib64 -rpath /nix/store/f1ilflm2wmkcq90b0bljcfs9vz0h41bb-shell/lib -L/nix/store/0vx5lr4639brlghdwn8fzgzvn7ca0hwp-zlib-1.2.11/lib -L/nix/store/0vx5lr4639brlghdwn8fzgzvn7ca0hwp-zlib-1.2.11/lib
```

Reply:
Right, libc requires further cc-wrapper simulation... I *was* thinking about libraries in nix-shells, yes. I assume that is how one would build Lake packages dependent on external libraries on NixOS.